README.md
#Aether Encoder IXGsw-v.01
This sw is develope for video transporting that is controlling by IGX-UI. 


ChangeLog.md
upto-30-12-2022 : Test json and gstremer pipeline test code.
                  Read and standing a CMake and create it. 
                  Create object for Json parsing, 
                  encoding pipeline, 
                  rtmp and monitor.
2jan-6jan2022   : Study about Dynamic tee pad. Read,Test functionallaty of connect and remove tee pad.


test/gstreamer/gstIpc.cpp
/*
livepipe = GST_PIPELINE (gst_parse_launch("interpipesrc listen-to=camsrc is-live=true allow-renegotiation=true stream-sync=restart-ts ! queue max-size-buffers=3 leaky=downstream"
      " ! nvvidconv ! video/x-raw(memory:NVMM),format=I420,width=1280,height=720 "
      " ! videorate ! video/x-raw(memory:NVMM),framerate=10/1 "
      " ! nvv4l2vp8enc bitrate=1000000 ! rtpvp8pay pt=96 ! udpsink host=127.0.0.1 port=5004", &error));
campipe = GST_PIPELINE (gst_parse_launch("nvarguscamerasrc ! video/x-raw(memory:NVMM),width=1932,height=1090,format=NV12,framerate=30/1 "
      " ! nvvidconv left=6 right=1926 top=6 bottom=1086 ! video/x-raw(memory:NVMM),width=1920,height=1080,format=RGBA "
      " ! nvvidconv ! video/x-raw(memory:NVMM),format=NV12 "
      " ! queue leaky=downstream max-size-buffers=3 "
      " ! interpipesink name=camsrc caps=video/x-raw(memory:NVMM),format=NV12 sync=true async=false", &error));

*/

#include <iostream>
#include <memory>
#include <thread>
#include <gst/gst.h>
#include <glib.h>

using namespace std;

#define GST_PLAY_KB_ARROW_UP    "\033[A"
#define GST_PLAY_KB_ARROW_DOWN  "\033[B"
#define GST_PLAY_KB_ARROW_RIGHT "\033[C"
#define GST_PLAY_KB_ARROW_LEFT  "\033[D"

class GSTElement{
    public:
        GMainLoop *loop1, *loop2, *loop3;
        GstElement *P1,*P2, *P3, *Src, *Vconv, *Vrate,*capsfilter, *capsfilterRate, *Queue;
        GstElement *Mux, *Sink, *H264Parse, *TimeOvlay, *Paktize, *capsfilterConvert;
        GstCaps *srccaps,*capsRate,*capsConvert;   
        GstBus *bus1, *bus2, *bus3;
        guint bus_watch_id;
        GError *error = NULL;
        GstStateChangeReturn ret;
        GIOChannel *io_stdin;

    public:
        int createEncodingPipe();
        int createRtmpPipe();
        int creteMulticatsPipe();
        int createDecodePipe();
};
/*
int GSTElement::createEncodingPipe()
{

}*/

void controlButton(GSTElement GE)
{
    cout << "i am thread" << endl;
    int i;

   // GSTElement GE;

    while(true)
    {
            cout << "waiting for key input" << endl;
            cin >> i;

            switch (i)
            {
            case 1:
                cout << "get key 1" << endl;
                gst_element_set_state (GE.P2, GST_STATE_PAUSED);
                break;
            case 2:
                cout << "get key 2" << endl;
                gst_element_set_state (GE.P2, GST_STATE_PLAYING);
                break;
            case 3:
                cout << "get key 1" << endl;
                gst_element_set_state (GE.P3, GST_STATE_PAUSED);
                break;
            case 4:
                cout << "get key 2" << endl;
                gst_element_set_state (GE.P3, GST_STATE_PLAYING);
                break;
            default:
                //gst_element_set_state (GE.P2, GST_STATE_NULL);
                break;
            }
    }
}

int main(int argc, char *argv[])
{
    GSTElement GE;
    gst_init (&argc, &argv);

    //GE.P1 = gst_parse_launch("videotestsrc ! video/x-raw,width=1920,height=1080,format=NV12,framerate=60000/1001 ! videoconvert ! video/x-raw,width=1920,height=1080,format=NV12 ! queue  ! interpipesink name=camsrc caps=video/x-raw,format=NV12 sync=true async=false", &GE.error);

   // GE.P2 = gst_parse_launch("interpipesrc listen-to=camsrc is-live=true allow-renegotiation=true stream-sync=restart-ts ! queue ! videoconvert ! video/x-raw,format=NV12,width=1920,height=1080 ! videorate ! video/x-raw,framerate=60000/1001 ! nvh264enc bitrate=10000 ! h264parse config-interval=-1 ! mpegtsmux name=mux alignment=7 ! udpsink host=226.1.1.1 port=23000", &GE.error);

    GE.P1 = gst_parse_launch("videotestsrc ! video/x-raw,width=1920,height=1080,format=NV12,framerate=60000/1001 ! videoconvert ! video/x-raw,width=1920,height=1080,format=NV12 ! videorate ! video/x-raw,framerate=60000/1001 ! nvh264enc bitrate=10000 ! h264parse config-interval=-1 ! mpegtsmux name=mux alignment=7 ! queue  ! interpipesink name=camsrc caps=video/x-raw,format=NV12 sync=true async=false", &GE.error);

    GE.P2 = gst_parse_launch("interpipesrc listen-to=camsrc is-live=true allow-renegotiation=true stream-sync=restart-ts ! udpsink host=226.1.1.1 port=23000", &GE.error);

    GE.P3 = gst_parse_launch("interpipesrc listen-to=camsrc is-live=true allow-renegotiation=true stream-sync=restart-ts ! udpsink host=226.1.1.1 port=23002", &GE.error);


   // g_signal_connect (GE.P1, "source-setup", G_CALLBACK (source_setup), &data);
    GE.io_stdin = g_io_channel_unix_new (fileno (stdin));
  
    /* Instruct the bus to emit signals for each received message, and connect to the interesting signals */
    GE.bus1 = gst_element_get_bus (GE.P1);
    GE.bus2 = gst_element_get_bus (GE.P2);
    GE.bus3 = gst_element_get_bus (GE.P3);
    gst_bus_add_signal_watch (GE.bus1);
    gst_bus_add_signal_watch (GE.bus2);
    gst_bus_add_signal_watch (GE.bus3);
   // g_signal_connect (G_OBJECT (bus), "message::error", (GCallback)error_cb, &data);
    gst_object_unref (GE.bus1);
    gst_object_unref (GE.bus2);
    gst_object_unref (GE.bus3);
    
    /* Start playing the pipeline */
    gst_element_set_state (GE.P1, GST_STATE_PLAYING);
    gst_element_set_state (GE.P2, GST_STATE_PLAYING);
    gst_element_set_state (GE.P3, GST_STATE_PLAYING);

    thread t1(controlButton,GE);
    t1.join();
    
    /* Create a GLib Main Loop and set it to run */
    GE.loop1 = g_main_loop_new (NULL, FALSE);
    GE.loop2 = g_main_loop_new (NULL, FALSE);
    GE.loop3 = g_main_loop_new (NULL, FALSE);
    g_main_loop_run (GE.loop1);
    g_main_loop_run (GE.loop2);
    g_main_loop_run (GE.loop3);
    
    /* Free resources */
    gst_element_set_state (GE.P1, GST_STATE_NULL);
    gst_element_set_state (GE.P2, GST_STATE_NULL);
    gst_element_set_state (GE.P3, GST_STATE_NULL);
    gst_object_unref (GE.P1);
    gst_object_unref (GE.P2);
    gst_object_unref (GE.P3);
    
}

test/json/jsoncpp_demo.cpp
#include <cstdlib>
#include <string>
#include <fstream>
#include <iostream>
#include <json/value.h>
#include <json/json.h>
using namespace std;
 
int main() {
    Json::Reader reader;  //for reading the data
    Json::Value root; //for modifying and storing new values
    Json::StyledStreamWriter writer; //for writing in json files
    ofstream newFile;
 
    //opening file using fstream
    ifstream file("sample.json");

 
    // check if there is any error is getting data from the json file
        if (!reader.parse(file, root)) {
        cout << reader.getFormattedErrorMessages();
        exit(1);
        }
       // cout << root << endl; // show the contain of file
        //Updating the json data
        cout << root["Input"][0]["type"]["SDI"]["active"] << endl;
 
        //we can add new values as well
       root["Input"][0]["type"]["SDI"]["active"] = "no";
      /*  root["last"] = "Jain";*/
 
        // make the changes in a new file/original file
        newFile.open("sample.json");
        writer.write(newFile, root);
        newFile.close();
    return 0;
}


test/gstreamer/gst_Ipc/gstIpc.cpp
/*
livepipe = GST_PIPELINE (gst_parse_launch("interpipesrc listen-to=camsrc is-live=true allow-renegotiation=true stream-sync=restart-ts ! queue max-size-buffers=3 leaky=downstream"
      " ! nvvidconv ! video/x-raw(memory:NVMM),format=I420,width=1280,height=720 "
      " ! videorate ! video/x-raw(memory:NVMM),framerate=10/1 "
      " ! nvv4l2vp8enc bitrate=1000000 ! rtpvp8pay pt=96 ! udpsink host=127.0.0.1 port=5004", &error));
campipe = GST_PIPELINE (gst_parse_launch("nvarguscamerasrc ! video/x-raw(memory:NVMM),width=1932,height=1090,format=NV12,framerate=30/1 "
      " ! nvvidconv left=6 right=1926 top=6 bottom=1086 ! video/x-raw(memory:NVMM),width=1920,height=1080,format=RGBA "
      " ! nvvidconv ! video/x-raw(memory:NVMM),format=NV12 "
      " ! queue leaky=downstream max-size-buffers=3 "
      " ! interpipesink name=camsrc caps=video/x-raw(memory:NVMM),format=NV12 sync=true async=false", &error));

*/

#include <iostream>
#include <memory>
#include <thread>
#include <gst/gst.h>
#include <glib.h>

using namespace std;

#define GST_PLAY_KB_ARROW_UP    "\033[A"
#define GST_PLAY_KB_ARROW_DOWN  "\033[B"
#define GST_PLAY_KB_ARROW_RIGHT "\033[C"
#define GST_PLAY_KB_ARROW_LEFT  "\033[D"

class GSTElement{
    public:
        GMainLoop *loop1, *loop2, *loop3;
        GstElement *P1,*P2, *P3, *Src, *Vconv, *Vrate,*capsfilter, *capsfilterRate, *Queue;
        GstElement *Mux, *Sink, *H264Parse, *TimeOvlay, *Paktize, *capsfilterConvert;
        GstCaps *srccaps,*capsRate,*capsConvert;   
        GstBus *bus1, *bus2, *bus3;
        guint bus_watch_id;
        GError *error = NULL;
        GstStateChangeReturn ret;
        GIOChannel *io_stdin;

    public:
        int createEncodingPipe();
        int createRtmpPipe();
        int creteMulticatsPipe();
        int createDecodePipe();
};
/*
int GSTElement::createEncodingPipe()
{

}*/

void controlButton(GSTElement GE)
{
    cout << "i am thread" << endl;
    int i;

   // GSTElement GE;

    while(true)
    {
            cout << "waiting for key input" << endl;
            cin >> i;

            switch (i)
            {
            case 1:
                cout << "get key 1" << endl;
                gst_element_set_state (GE.P2, GST_STATE_PAUSED);
                break;
            case 2:
                cout << "get key 2" << endl;
                gst_element_set_state (GE.P2, GST_STATE_PLAYING);
                break;
            case 3:
                cout << "get key 1" << endl;
                gst_element_set_state (GE.P3, GST_STATE_PAUSED);
                break;
            case 4:
                cout << "get key 2" << endl;
                gst_element_set_state (GE.P3, GST_STATE_PLAYING);
                break;
            default:
                //gst_element_set_state (GE.P2, GST_STATE_NULL);
                break;
            }
    }
}

int main(int argc, char *argv[])
{
    GSTElement GE;
    gst_init (&argc, &argv);

    //GE.P1 = gst_parse_launch("videotestsrc ! video/x-raw,width=1920,height=1080,format=NV12,framerate=60000/1001 ! videoconvert ! video/x-raw,width=1920,height=1080,format=NV12 ! queue  ! interpipesink name=camsrc caps=video/x-raw,format=NV12 sync=true async=false", &GE.error);

   // GE.P2 = gst_parse_launch("interpipesrc listen-to=camsrc is-live=true allow-renegotiation=true stream-sync=restart-ts ! queue ! videoconvert ! video/x-raw,format=NV12,width=1920,height=1080 ! videorate ! video/x-raw,framerate=60000/1001 ! nvh264enc bitrate=10000 ! h264parse config-interval=-1 ! mpegtsmux name=mux alignment=7 ! udpsink host=226.1.1.1 port=23000", &GE.error);

    GE.P1 = gst_parse_launch("videotestsrc ! video/x-raw,width=1920,height=1080,format=NV12,framerate=60000/1001 ! videoconvert ! video/x-raw,width=1920,height=1080,format=NV12 ! videorate ! video/x-raw,framerate=60000/1001 ! nvh264enc bitrate=10000 ! h264parse config-interval=-1 ! mpegtsmux name=mux alignment=7 ! queue ! interpipesink name=camsrc sync=true async=false", &GE.error);

    GE.P2 = gst_parse_launch("interpipesrc listen-to=camsrc is-live=true allow-renegotiation=true stream-sync=restart-ts ! udpsink host=226.1.1.1 port=23000", &GE.error);

    GE.P3 = gst_parse_launch("interpipesrc listen-to=camsrc is-live=true allow-renegotiation=true stream-sync=restart-ts ! udpsink host=226.1.1.1 port=23002", &GE.error);


   // g_signal_connect (GE.P1, "source-setup", G_CALLBACK (source_setup), &data);
    GE.io_stdin = g_io_channel_unix_new (fileno (stdin));
  
    /* Instruct the bus to emit signals for each received message, and connect to the interesting signals */
    GE.bus1 = gst_element_get_bus (GE.P1);
    GE.bus2 = gst_element_get_bus (GE.P2);
    GE.bus3 = gst_element_get_bus (GE.P3);
    gst_bus_add_signal_watch (GE.bus1);
    gst_bus_add_signal_watch (GE.bus2);
    gst_bus_add_signal_watch (GE.bus3);
   // g_signal_connect (G_OBJECT (bus), "message::error", (GCallback)error_cb, &data);
   // gst_object_unref (GE.bus1);
   // gst_object_unref (GE.bus2);
   // gst_object_unref (GE.bus3);
    
    /* Start playing the pipeline */
    gst_element_set_state (GE.P1, GST_STATE_PLAYING);
    gst_element_set_state (GE.P2, GST_STATE_PLAYING);
    gst_element_set_state (GE.P3, GST_STATE_PLAYING);

    thread t1(controlButton,GE);
    t1.join();
    
    /* Create a GLib Main Loop and set it to run */
    GE.loop1 = g_main_loop_new (NULL, FALSE);
    GE.loop2 = g_main_loop_new (NULL, FALSE);
    GE.loop3 = g_main_loop_new (NULL, FALSE);
    cout<<"run loop 1"<<endl;
    g_main_loop_run (GE.loop1);
    cout<<"run loop 2"<<endl;
    g_main_loop_run (GE.loop2);
    cout<<"run loop 3"<<endl;
    g_main_loop_run (GE.loop3);
    cout<<"run all loop"<<endl;
    
    /* Free resources */
    gst_element_set_state (GE.P1, GST_STATE_NULL);
    gst_element_set_state (GE.P2, GST_STATE_NULL);
    gst_element_set_state (GE.P3, GST_STATE_NULL);
    gst_object_unref (GE.P1);
    gst_object_unref (GE.P2);
    gst_object_unref (GE.P3);
    
}

src/gstreamer/gst.h
#ifndef GST_INCLUDE
#define GST_INCLUDE

#include <memory>
#include <gst/gst.h>
#include <glib.h>
#include "../json/jsonParser.h"

class gstEncodeElement{
    public:
        GMainLoop *loop;
        GstElement *P, *src, *vconv, *vrate,*capsfiltersrc, *capsfilterRate, *queue, *encoder, *muxer, *filesink, *videoconvert, *videosink, *queue_record, *parse;
        GstElement *mux, *sink, *H264Encode, *H264Parse, *capsfilterConvert, *capsfilterH264enc;
        GstCaps *srccaps,*capsRate,*capsConvert, *capsH264enc;   
        GstBus *bus;
        guint bus_watch_id;
        GError *error = NULL;
        GstStateChangeReturn ret;
        GIOChannel *io_stdin;
        GstState state;
        GstPad *teepad;
        gboolean recording = FALSE;

    public:
        int createEncodePipe(AetherAction AA);
        int freeEncodePipe();
        void startRecording();
        void stopRecording();
};

class createMulticastPipe{
   
    public:
        createMulticastPipe(AetherAction AA);
};

class createRtmpPipe{
     public:
        GMainLoop *loop1;
        GstElement *P1;
        GstBus *bus1;
        GError *error = NULL;
        GstState state;
    public:
        int rtmpPipe(string rtmplink, string node);
        void freeRtmpPipe();       
};


#endif

src/AetherEncoder.cpp
#include "AetherEncoder.h"
#include "gstreamer/gst.h"
#include "json/jsonParser.h"
#include <thread>
#include <signal.h>

using namespace std;

int handleYoutubePipeline(struct jsonParameter *YJP, createRtmpPipe *newRtmpPipe, int ch)
{
   // int ch;
    
    if(YJP[0].ytb[ch].active == "start"){
        //createRtmpPipe *newRtmpPipe1 = new createRtmpPipe;
        newRtmpPipe[ch].rtmpPipe(YJP[0].ytb[ch].link, YJP[0].ytb[ch].streamto);
    }
    else if(YJP[0].ytb[ch].active == "stop") {
        newRtmpPipe[YJP[0].ytb[ch].youtubeNO].freeRtmpPipe();
                //delete &newRtmpPipe[YJP.ytb.youtubeNO];
    }

    return 0;
}
/*
int sigintHandler(gstEncodeElement GE) {
	g_print("You ctrl-c!\n");
	if (GE.recording)
		GE.stopRecording();
	else
		GE.startRecoding();
	return 0;
}
*/
int main(int argc, char *argv[])
{
    gstEncodeElement GE;
    createRtmpPipe RP;
    AetherAction AA;
    struct jsonParameter *JP = new jsonParameter[8];
    createRtmpPipe *newRtmpPipe = new createRtmpPipe[8];
    int ytbch;
    //signal(SIGINT, sigintHandler);
    cout<< "Initialization gstreamer\n" << endl;
    gst_init (&argc, &argv);
    
    //parse json file
    JP[0]=AA.parseJson("../sample.json",JP);

    cout<< "setup encoding pipe" << endl;
    GE.createEncodePipe(AA);

    while(true)
    {
        string filePath="../cmd/ytb.json";
        g_usleep(1000000);
        if(AA.fileExist(filePath)){
            cout << "call youtube json parser" << endl;
            JP[0] = AA.parseYtbJson(filePath, JP, &ytbch);
            handleYoutubePipeline(JP,newRtmpPipe,ytbch); 
            remove(filePath.c_str());
        }

      /*  if (GE.recording){
            g_print("stop recording\n");
		  //  GE.stopRecording();
        }
	    else{
            g_print("start recording\n");
		  //  GE.startRecording();
        }*/
    }

    //RP.freeRtmpPipe();
    GE.freeEncodePipe();   
}

src/json/jsonParser.h
#ifndef JSONPARSER_INCLUDE
#define JSONPARSER_INCLUDE

using namespace std;

struct resolution{
    int height;
    int width;
};

struct IP{
    string ipAddr;
    int port;
    string active;
    string input_type;
    string sys_IP;
};

struct SDI{
     string input_type;
     string active;
     int sdi_num;
     string dev_name;
};

struct TYPE{
    struct IP ip;
    struct SDI sdi;
};

struct input{
    string active_type;
    struct resolution res;
    float framerate; 
    struct TYPE type; 
};

struct output{
    string active_type;
    int bitrate;
    string encode;
    float framerate;
    struct resolution res;
    struct TYPE type; 
};

struct youtube{
    string streamto;		
	string active;
	string key ;
	string link;
    int youtubeNO;
};

struct jsonParameter{
    struct input in;
    struct output out;
    struct youtube ytb[10]; 
};

class AetherAction{
    public:
        Json::Reader reader;  //for reading the data
        Json::Value root; //for modifying and storing new values
        Json::StyledStreamWriter writer; //for writing in json files
        struct jsonParameter JP[10];
       // bool status;
        
        
    public:
        jsonParameter parseJson(string fpath, jsonParameter *JP);
        int updateJson(string path);
        int showJsonContain(string path);
        string showJsonValueStr(string key, string value);
        int showJsonValueInt(string key, int value);
        bool fileExist(string filePath);
        jsonParameter parseYtbJson(string fpath, jsonParameter *JP, int* ch);
};

#endif

src/json/jsonParser.cpp
#include "../AetherEncoder.h"
#include "jsonParser.h"
#include <unistd.h>


using namespace std;
 
  struct jsonParameter AetherAction::parseJson(string fpath,  struct jsonParameter *JP) 
 {
    //ofstream newJson;
   // struct input in;
   // struct output out;
   // struct jsonParameter jsonParam;
    cout << fpath << endl;
    //opening file using fstream
    ifstream json(fpath);
    if(!json.is_open())
    {
        cout<<"File doesn't exits" << endl;
        exit(0);
    }
 
    // check if there is any error is getting data from the json file
    if (!reader.parse(json, root)) {
        cout << reader.getFormattedErrorMessages();
        exit(1);
    }
    
    JP[0].in.active_type = root["Input"][0]["active_type"].asString();
    JP[0].in.framerate = root["Input"][0]["framerate"].asFloat();
    JP[0].in.res.height = root["Input"][0]["resolution"]["height"].asInt();
    JP[0].in.res.width = root["Input"][0]["resolution"]["width"].asInt();
    JP[0].in.type.ip.active = root["Input"][0]["type"]["IP"]["active"].asString();
    JP[0].in.type.ip.input_type = root["Input"][0]["type"]["IP"]["input_type"].asString();
    JP[0].in.type.ip.ipAddr = root["Input"][0]["type"]["IP"]["ipAddr"].asString();
    JP[0].in.type.ip.port = root["Input"][0]["type"]["IP"]["port"].asInt();
    JP[0].in.type.ip.sys_IP = root["Input"][0]["type"]["IP"]["input_type"].asString();

    JP[0].out.active_type = root["Input"][0]["active_type"].asString();
    JP[0].out.bitrate = root["Input"][0]["bitrate"].asInt();
    JP[0].out.encode = root["Input"][0]["encode"].asString();
    JP[0].out.framerate = root["Input"][0]["framerate"].asFloat();
    JP[0].out.res.height = root["Input"][0]["resolution"]["height"].asInt();
    JP[0].out.res.width = root["Input"][0]["resolution"]["weight"].asInt();
    JP[0].out.type.ip.active = root["Input"][0]["type"]["IP"]["active"].asString();
    JP[0].out.type.ip.input_type = root["Input"][0]["type"]["IP"]["input_type"].asString();
    JP[0].out.type.ip.ipAddr = root["Input"][0]["type"]["IP"]["ipAddr"].asString();
    JP[0].out.type.ip.port = root["Input"][0]["type"]["IP"]["port"].asInt();
    JP[0].out.type.ip.sys_IP = root["Input"][0]["type"]["IP"]["input_type"].asString();

    //cout << out.active_type << endl << out.framerate << endl << out.res.height << endl << out.type.ip.active << endl;
    //cout << JP[0].in.res.width << endl;
    //make the changes in a new file/original file
    //newJson.open(fpath);
    //writer.write(newJson, root);
    //newJson.close();
    return JP[0];
}
bool AetherAction::fileExist(string filePath)
{
   return ( access( filePath.c_str(), F_OK ) != -1 ); 
}

struct jsonParameter AetherAction::parseYtbJson(string fpath, jsonParameter *JP, int* ch)
{
    struct jsonParameter ytbParam;
    //opening file using fstream
    ifstream json(fpath);
    int nch;

    if(!json.is_open())
        cout<<"cmd file doesnit exits"<<endl;
    // check if there is any error is getting data from the json file
    if (!reader.parse(json, root)) {
        cout << reader.getFormattedErrorMessages();
        exit(1);
    }
    *ch = root["youtubeNO"].asInt();
    nch = *ch;
    JP[0].ytb[nch].active = root["active"].asString();
    JP[0].ytb[nch].youtubeNO = root["youtubeNO"].asInt();
    JP[0].ytb[nch].streamto = root["streamto"].asString();
    JP[0].ytb[nch].key = root["key"].asString();
    JP[0].ytb[nch].link = root["link"].asString();

    cout << JP[0].ytb[nch].active << endl;
    cout << JP[0].ytb[nch].youtubeNO << endl;
    cout << JP[0].ytb[nch].streamto << endl;
    cout << JP[0].ytb[nch].key << endl;
    cout << JP[0].ytb[nch].link << endl;

    return JP[0];
}

src/gstreamer/gst.cpp
/*
gst-launch-1.0 videotestsrc ! video/x-raw,width=1920,height=1080,format=NV12,framerate=60000/1001 ! videoconvert ! video/x-raw,width=1920,height=1080,format=NV12 ! 
queue leaky=downstream max-size-buffers=3 ! interpipesink name=camsrc caps=video/x-raw,format=NV12 sync=true async=false

gst-launch-1.0 decklinkvideosrc device-number=0 connection=sdi mode=1080p5994 buffer-size=5 ! "video/x-raw,format=UYVY" ! videoconvert ! "video/x-raw,format=NV12" ! 
nvh264enc  zerolatency=true strict-gop=false gop-size=60 bframes=0  nonref-p=false  rc-lookahead=0 qp-const=0 rc-mode=2 preset=4 vbv-buffer-size=1024 bitrate=8000 max-bitrate=8000 ! 
"video/x-h264,profile=(string)main" ! h264parse config-interval=1 ! mpegtsmux alignment=7 name=mux decklinkaudiosrc device-number=0 channels=2 buffer-size=5 ! "audio/x-raw,format=S16LE" ! 
audioconvert ! voaacenc bitrate=128000 ! aacparse ! mux. mux. ! tee name=t ! queue ! srtsink streamid=1 uri=srt://104.189.184.146:23002 wait-for-connection=false  sync=true
*/

#include "../AetherEncoder.h"
#include "gst.h"


//static gboolean bus_call (GstBus *bus, GstMessage *msg, gpointer data);

static gboolean
message_cb (GstBus * bus, GstMessage * message, gpointer data)
{
  switch (GST_MESSAGE_TYPE (message)) {
    case GST_MESSAGE_ERROR:{
      GError *err = NULL;
      gchar *name, *debug = NULL;

      name = gst_object_get_path_string (message->src);
      gst_message_parse_error (message, &err, &debug);

      g_printerr ("ERROR: from element %s: %s\n", name, err->message);
      if (debug != NULL)
        g_printerr ("Additional debug info:\n%s\n", debug);

      g_error_free (err);
      g_free (debug);
      g_free (name);

     // g_main_loop_quit (data.loop);
      break;
    }
    case GST_MESSAGE_WARNING:{
		GError *err = NULL;
		gchar *name, *debug = NULL;

		name = gst_object_get_path_string (message->src);
		gst_message_parse_warning (message, &err, &debug);

		g_printerr ("ERROR: from element %s: %s\n", name, err->message);
		if (debug != NULL)
		g_printerr ("Additional debug info:\n%s\n", debug);

		g_error_free (err);
		g_free (debug);
		g_free (name);
		break;
    }
    case GST_MESSAGE_EOS: {
		g_print ("Got EOS\n");
		// g_main_loop_quit (loop);
		// gst_element_set_state (pipeline, GST_STATE_NULL);
		// g_main_loop_unref (loop);
		// gst_object_unref (pipeline);
		// exit(0);
		break;
	}
	case GST_MESSAGE_ELEMENT: {
    	const GstStructure *s = gst_message_get_structure (message);

    	if (gst_structure_has_name (s, "GstBinForwarded")) {
	        GstMessage *forward_msg = NULL;

	        gst_structure_get (s, "message", GST_TYPE_MESSAGE, &forward_msg, NULL);
	        if (GST_MESSAGE_TYPE (forward_msg) == GST_MESSAGE_EOS) {
	            g_print ("EOS from element %s\n", GST_OBJECT_NAME (GST_MESSAGE_SRC (forward_msg)));	
	           // finalize();
	        }
	        gst_message_unref (forward_msg);
    	}
    	break;
	}
    default:
		break;
  }

  return TRUE;
}

int gstEncodeElement::createEncodePipe(AetherAction AA)
{
    
    //string pipelineStr = "decklinkvideosrc device-number=0 connection=sdi mode=1080p5994 buffer-size=5 ! video/x-raw,format=UYVY,height=1080,width=1920 ! videoconvert ! video/x-raw,format=NV12,width=1920,height=1080 ! nvh264enc  zerolatency=true strict-gop=false gop-size=120 bframes=0  nonref-p=false  rc-lookahead=0 rc-mode=1 preset=4 vbv-buffer-size=1024 bitrate=8000 max-bitrate=8000 ! video/x-h264,profile=(string)main ! h264parse config-interval=-1 ! flvmux streamable=true name=mux decklinkaudiosrc device-number=0 channels=2 buffer-size=5 ! audio/x-raw,format=S16LE ! audioconvert ! voaacenc bitrate=128000 ! aacparse ! mux. mux. ! queue ! interpipesink name=camsrc sync=true async=false";
    string pipelineStr = "decklinkvideosrc device-number=0 connection=sdi mode=1080p5994 buffer-size=5 ! video/x-raw,width=1920,height=1080 ! videoconvert ! video/x-raw,format=NV12 ! tee name=vraw ! nvh264enc  zerolatency=true strict-gop=false gop-size=120 bframes=0  nonref-p=false  rc-lookahead=0 rc-mode=1 preset=4 vbv-buffer-size=1024 bitrate=8000 max-bitrate=8000 ! video/x-h264,profile=main ! h264parse config-interval=-1 ! flvmux streamable=true name=mux ! interpipesink name=rtmp sync=true async=false decklinkaudiosrc device-number=0 channels=2 buffer-size=5 ! audio/x-raw,format=S16LE ! audioconvert ! voaacenc bitrate=128000 ! tee name=at ! aacparse ! mux. vraw. ! queue ! videorate ! video/x-raw,framerate=30000/1001 ! videoscale ! video/x-raw,width=320,height=240 ! nvh264enc  zerolatency=true strict-gop=false gop-size=60 bframes=0  nonref-p=false  rc-lookahead=0 rc-mode=1 preset=4 vbv-buffer-size=512 bitrate=500 max-bitrate=600 ! video/x-h264,profile=baseline ! h264parse config-interval=-1 ! flvmux streamable=true name=mmux ! interpipesink name=monitor sync=true async=false at. ! queue ! aacparse ! mmux.";
    P = gst_parse_launch(pipelineStr.c_str(), &error);
    if (error) {
    g_print ("Error while parsing pipeline description: %s\n", error->message);
    return -1;
    }
    g_object_set(GST_BIN(P), "message-forward", TRUE, NULL);
    /* Instruct the bus to emit signals for each received message, and connect to the interesting signals */
    bus = gst_element_get_bus (P);   
    gst_bus_add_signal_watch (bus);
    
    /* Start playing the pipeline */
    gst_element_set_state (P, GST_STATE_PLAYING);

    /* Create a GLib Main Loop and set it to run */
    loop = g_main_loop_new (NULL, FALSE);
   // g_signal_connect(G_OBJECT(bus), "message", G_CALLBACK(message_cb), gstEncodeElement GE, NULL);
   // g_main_loop_run (loop);
    if (gst_element_get_state (P, &state, NULL, GST_CLOCK_TIME_NONE) == GST_STATE_CHANGE_FAILURE || state != GST_STATE_PLAYING) {
    g_warning ("State change to playing failed");
    }   
    return 0;
}

int gstEncodeElement::freeEncodePipe()
{
    /* Free resources */
    gst_element_set_state (P, GST_STATE_NULL);
    gst_object_unref (P);
    return 0;
}

int createRtmpPipe::rtmpPipe(string rtmplink, string node)
{
    cout << "set rtmp pipe" << endl;
    //string rtmpStr = "interpipesrc listen-to=" + node + " is-live=true allow-renegotiation=true stream-sync=restart-ts ! queue ! rtmp2sink location=\"rtmp://a.rtmp.youtube.com/live2/x/" + rtmpKey +"\"";
    string rtmpStr = "interpipesrc listen-to=" + node + " is-live=true allow-renegotiation=true stream-sync=restart-ts ! queue ! rtmp2sink location=\"" + rtmplink + "\"";
    cout << rtmpStr << endl;
    P1 = gst_parse_launch(rtmpStr.c_str(), &error);

    if (error) {
    g_print ("Error while parsing pipeline description: %s\n", error->message);
    return -1;
    }

    loop1 = g_main_loop_new (NULL, FALSE);
    bus1 = gst_element_get_bus (P1);
    gst_bus_add_signal_watch (bus1);
    gst_element_set_state (P1, GST_STATE_PLAYING);
    /* lets check it gets to PLAYING */
    if (gst_element_get_state (P1, &state, NULL, GST_CLOCK_TIME_NONE) == GST_STATE_CHANGE_FAILURE || state != GST_STATE_PLAYING) {
    g_warning ("State change to playing failed");
    }
    
    return 0;
}
void createRtmpPipe::freeRtmpPipe()
{   
    cout << "free rtmp pipe" << endl;
  //  g_main_loop_run (loop1);
    gst_element_set_state(P1, GST_STATE_PAUSED);
    g_usleep(100);
    gst_element_set_state (P1, GST_STATE_NULL);
    gst_object_unref (P1);
    gst_object_unref (bus1);
}